ProjFS.Mac/PrjFSKextTests/MockVnodeAndMount.cpp

```diff
@@ -0,0 +1,219 @@
+{
+    VnodePointer result(new vnode(mount));
+    s_allVnodes.insert(make_pair(result.get(), VnodeWeakPointer(result)));
+    result->weakSelfPointer = result;
+    result->SetPath(path);
+    result->type = vnodeType;
+    return result;
+}
+
+VnodePointer vnode::Create(const MountPointer& mount, const char* path, vtype vnodeType, uint64_t inode)
+{
+    VnodePointer result = Create(mount, path, vnodeType);
+    result->inode = inode;
+    return result;
+}
+
+void vnode::StartRecycling()
+{
+    s_vnodesByPath.erase(this->path);
+    this->path.clear();
+    this->type = VBAD;
+    this->vid++;
+    this->isRecycling = true;
+}
+
+void vnode::SetGetPathError(errno_t error)
+{
+    this->getPathError = error;
+}
+
+void vnode::SetPath(const string& path)
+{
+    s_vnodesByPath.erase(this->path);
+
+    this->path = path;
+    size_t lastSlash = this->path.rfind('/');
+    if (lastSlash == string::npos)
+    {
+        this->name = this->path.c_str();
+    }
+    else
+    {
+        this->name = this->path.c_str() + lastSlash + 1;
+    }
+    
+    assert(!s_vnodesByPath[path].lock());
+    s_vnodesByPath[path] = this->weakSelfPointer;
+}
+
+int vnode_isrecycled(vnode_t vnode)
+{
+    return vnode->isRecycling;
+}
+
+const char* vnode_getname(vnode_t vnode)
+{
+    return vnode->name;
+}
+
+void vnode_putname(const char* name)
+{
+    // TODO: track name reference counts
+}
+
+int vnode_isdir(vnode_t vnode)
+{
+    return vnode_vtype(vnode) == VDIR;
+}
+
+int vn_getpath(vnode_t vnode, char* pathBuffer, int* pathLengthInOut)
+{
+    if (vnode->getPathError != 0)
+    {
+        return vnode->getPathError;
+    }
+    else if (vnode->path.empty() || vnode->isRecycling)
+    {
+        // TODO: check what the real vn_getpath() would return here
+        return EIO;
+    }
+    else
+    {
+        strlcpy(pathBuffer, vnode->path.c_str(), MIN(*pathLengthInOut, MAXPATHLEN));
+        return 0;
+    }
+}
+
+// TODO: Perhaps switch to the real version of this function and mock the KPIs it uses
+FsidInode Vnode_GetFsidAndInode(vnode_t vnode, vfs_context_t vfsContext)
+{
+    return FsidInode{ vnode->mountPoint->statfs.f_fsid, vnode->inode };
+}
+
+errno_t vnode_lookup(const char* path, int flags, vnode_t* foundVnode, vfs_context_t vfsContext)
+{
+    PathToVnodeMap::const_iterator found = s_vnodesByPath.find(path);
+    if (found == s_vnodesByPath.end())
+    {
+        return ENOENT;
+    }
+    else if (VnodePointer vnode = found->second.lock())
+    {
+        // vnode_lookup returns a vnode with an iocount
+        ++vnode->ioCount;
+        *foundVnode = vnode.get();
+        return 0;
+    }
+    else
+    {
+        s_vnodesByPath.erase(found);
+        return ENOENT;
+    }
+}
+
+uint32_t vnode_vid(vnode_t vnode)
+{
+    return vnode->vid;
+}
+
+vtype vnode_vtype(vnode_t vnode)
+{
+    return vnode->type;
+}
+
+mount_t vnode_mount(vnode_t vnode)
+{
+    return vnode->mountPoint.get();
+}
+
+int vnode_put(vnode_t vnode)
```

Comment:
Asking to confirm my understanding of put/get.
put - removes
get - inserts

If we have a get unpaired with a put, what could happen?
if we have put without a get, what could happen?

Reply:
`vnode_get` and `vnode_put` increment and decrement the iocount in the vnode. This roughly corresponds to opening and closing file descriptors.

vnodes can only become "dead" (`VBAD`) if the iocount is 0. This is a prerequisite for recycling, so if we were missing a `vnode_put`, we'd essentially be leaking vnodes. Worse, file system mount points (OS ones, not VFS4G ones) can only be unmounted if all vnodes for files on them are dead. So that prevents ejecting external volumes, etc.

If you call `vnode_put()` on a vnode with an iocount of 0, that causes a kernel panic, see #797.

Reply:
@jrbriggs this is a very elegant explanation of the get/put requirement.